### PR TITLE
Add support for SAML authentication

### PR DESCRIPTION
**Pull request by EfrenRey**
@davisein @manolakis 

@mattnickles Can you review saml settings design?

### CODE REVIEW

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/ek_pull_request/10/ for more details.

**Comment by EfrenRey** [comment]
Jenkins, test this please

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/ek_pull_request/12/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/ek_pull_request/16/ for more details.

**Comment by albertoarias** [comment]
@EfrenRey Please rebase

**Comment by albertoarias** [comment]
@EfrenRey Please rebase so we can get the test instance

**Comment by allyourbotsarebelongtous** [comment]
Instance i-yc2iv0 couldnt be deleted. It requires manual deletion

**Comment by allyourbotsarebelongtous** [comment]
Instance i-7itcnz couldnt be deleted. It requires manual deletion

**Comment by allyourbotsarebelongtous** [comment]
Instance i-lde786 couldnt be deleted. It requires manual deletion

**Comment by allyourbotsarebelongtous** [comment]
The following instances are being terminated: https://elasticbox.com/services/instances/i-yc2iv0, https://elasticbox.com/services/instances/i-7itcnz, https://elasticbox.com/services/instances/i-lde786

**Comment by allyourbotsarebelongtous** [comment]
Deployed test instance is available at http://ek-43.internal.elastickube.com

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/ek_pull_request/49/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/ek_pull_request/50/ for more details.

**Comment by albertoarias** [comment]
- The radio button click does not work when clicking on the text (only if I click in the circle)
- Cannot upload the file

**Comment by arnaud-elasticbox** [comment]
screen shot 2016-05-12 at 9 08 34 am

**Comment by arnaud-elasticbox** [comment]
I got an error when I logged in the first time.

**Comment by albertoarias** [comment]
Jenkins test this please.

**Comment by allyourbotsarebelongtous** [comment]
Deployed test instance is available at http://ek-43.internal.elastickube.com

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/ek_pull_request/54/ for more details.

**Comment by arnaud-elasticbox** [comment]
@EfrenRey pep8 test is failing.

**Comment by EfrenRey** [comment]
Jenkins, test this please.

**Comment by allyourbotsarebelongtous** [comment]
Deployed test instance is available at http://ek-43.internal.elastickube.com

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/ek_pull_request/77/ for more details.

File: src/ui/app/admin/ek-admin-settings/ek-admin-settings.html [hunk 1]
```diff
@@ -33,53 +33,88 @@ <h4 class="ek-admin-settings__title">cluster administrators</h4>
...
+                                <p class="ek-admin-settings__authentication__methods__configuration__file--provider">Provider domain</p>
+                                <p>{{ ctrl.auth.saml_data.idp_domain }}</p>
+                            </div>    
+                        </div>
+                        
+                        <md-button id="uploadMetadataButton" class="ek-admin-settings__button--file ek-button"
+                                   ng-click="ctrl.uploadMetadata()">Update</md-button>
+                    </div>
+
+                    <div class="ek-message-error" ng-show="ctrl.metadata_error">
```
Comment (manolakis) [inline src/ui/app/admin/ek-admin-settings/ek-admin-settings.html]: Please, change the `ng-show` by `ng-if`

File: src/ui/app/admin/ek-admin-settings/ek-admin-settings.controller.js [hunk 2]
```diff
@@ -86,12 +111,24 @@ class AdminSettingsController {
...
+        if (this.auth.google && hasValues(this.auth.google_data)) {
+            this.auth_sso = 'google';
+        } else if (this.auth.saml && hasValues(this.auth.saml_data)) {
+            this.auth_sso = 'saml';
+        } else {
+            this.auth_sso = 'off';
+        }
+
+        if (this.auth_sso !== 'saml') {
+            this.metadata_error = _.undefined;
```
Comment (manolakis) [inline src/ui/app/admin/ek-admin-settings/ek-admin-settings.controller.js]: We disallow the use of `undefined` and `_.undefined` seems to be a trick to bypass it ;)

Please use `delete this.metadata_error` instead.

File: src/ui/app/admin/ek-admin-settings/ek-admin-settings.controller.js [hunk 3]
```diff
@@ -138,21 +178,17 @@ class AdminSettingsController {
             }
 
             if (!_.isEqual(this._settings, this._settingsStore.getSettings())) {
-                this._settingsActionCreator.update(this._settings);
+                this._settingsActionCreator
+                    .update(this._settings)
+                    .catch((error) => {
+                        this._$log.error(error.body);
+                        this._settingsActionCreator.unsubscribe();
+                        this._settingsActionCreator.subscribe();
```
Comment (manolakis) [inline src/ui/app/admin/ek-admin-settings/ek-admin-settings.controller.js]: Why is this needed?

File: src/ui/app/admin/ek-admin-settings/ek-admin-settings.controller.js [hunk 4]
```diff
@@ -65,6 +68,28 @@ class AdminSettingsController {
...
+                this._$scope.$evalAsync(() => {
+                    delete this._settings.authentication.google_oauth;
+                    if (!this._settings.authentication.saml) {
+                        this._settings.authentication.saml = {};
+                    }
+                    this._settings.authentication.saml.metadata = e.target.result;
+
+                    this._settingsActionCreator
+                        .update(this._settings)
+                        .then(() => this.metadata_error = _.undefined)
```
Comment (manolakis) [inline src/ui/app/admin/ek-admin-settings/ek-admin-settings.controller.js]: Please, use `delete this.metadata_error` instead

File: src/ui/app/admin/ek-admin-settings/ek-admin-settings.html [hunk 5]
```diff
@@ -33,53 +33,88 @@ <h4 class="ek-admin-settings__title">cluster administrators</h4>
...
+                        {{ ctrl.metadata_error }}
```
Comment (manolakis) [inline src/ui/app/admin/ek-admin-settings/ek-admin-settings.html]: Use a one time binding also with the `ng-if`
```
{{:: ctrl.metadata_error }}
```

File: src/ui/app/admin/ek-admin-settings/ek-admin-settings.less [hunk 6]
```diff
@@ -35,6 +35,49 @@
...
+                }
+            }
+            &__configuration {
+                padding: 0 0 20px 36px;
+                
+                &__file {
+                    border: 1px solid @grey300;
+                    box-shadow: 0 0 3px rgba(0, 0, 0, 0.08);
+                    margin-bottom: 10px;
+                    margin-top: 10px;
```
Comment (manolakis) [inline src/ui/app/admin/ek-admin-settings/ek-admin-settings.less]: You can use the compact version
``` css
margin: 10px 0;
```

File: src/ui/app/core/directives/file-change.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        super();
+
+        this.scope = false;
+        this.restrict = 'A';
+    }
+
+    link($scope, $element, attrs) {
+        const onChangeHandler = $scope.$eval(attrs.ekFileChange);
+
+        $scope.$evalAsync(() => $element.bind('change', onChangeHandler));
```
Comment (manolakis) [inline src/ui/app/core/directives/file-change.js]: Try this
```javascript
$element.bind('change', () => $scope.$apply(onChangeHandler));
```

File: src/ui/app/core/directives/file-change.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import Directive from 'directive';
+
+class FilechangeDirective extends Directive {
```
Comment (manolakis) [inline src/ui/app/core/directives/file-change.js]: OnFileChange ?

Comment (EfrenRey) [inline src/ui/app/core/directives/file-change.js]: Done

File: src/ui/app/core/directives/file-change.js (hunk 7 above)
Comment (EfrenRey) [inline src/ui/app/core/directives/file-change.js]: Done. Added parameter 'event'

File: src/ui/app/admin/ek-admin-settings/ek-admin-settings.less (hunk 6 above)
Comment (EfrenRey) [inline src/ui/app/admin/ek-admin-settings/ek-admin-settings.less]: Done

File: src/ui/app/admin/ek-admin-settings/ek-admin-settings.html (hunk 5 above)
Comment (EfrenRey) [inline src/ui/app/admin/ek-admin-settings/ek-admin-settings.html]: Done

File: src/ui/app/admin/ek-admin-settings/ek-admin-settings.controller.js (hunk 4 above)
Comment (EfrenRey) [inline src/ui/app/admin/ek-admin-settings/ek-admin-settings.controller.js]: Done

File: src/ui/app/admin/ek-admin-settings/ek-admin-settings.controller.js (hunk 2 above)
Comment (EfrenRey) [inline src/ui/app/admin/ek-admin-settings/ek-admin-settings.controller.js]: Done

File: src/ui/app/admin/ek-admin-settings/ek-admin-settings.html (hunk 1 above)
Comment (EfrenRey) [inline src/ui/app/admin/ek-admin-settings/ek-admin-settings.html]: Done

File: src/ui/app/admin/ek-admin-settings/ek-admin-settings.controller.js (hunk 3 above)
Comment (EfrenRey) [inline src/ui/app/admin/ek-admin-settings/ek-admin-settings.controller.js]: Removed

**Comment by allyourbotsarebelongtous** [comment]
Deployed test instance is available at http://ek-43.internal.elastickube.com

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/ek_pull_request/82/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Deployed test instance is available at http://ek-43.internal.elastickube.com

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/ek_pull_request/83/ for more details.

**Comment by davisein** [comment]
ACK

**Comment by allyourbotsarebelongtous** [comment]
The following instances are being terminated: https://elasticbox.com/services/instances/i-un9nl4

**Comment by allyourbotsarebelongtous** [comment]
Deployed test instance is available at http://ek-43.internal.elastickube.com

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/ek_pull_request/84/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Deployed test instance is available at http://ek-43.internal.elastickube.com

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/ek_pull_request/89/ for more details.

**Comment by EfrenRey** [comment]
Jenkins, test this please.

**Comment by allyourbotsarebelongtous** [comment]
Deployed test instance is available at http://ek-43.internal.elastickube.com

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/ek_pull_request/91/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Deployed test instance is available at http://ek-43.internal.elastickube.com

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/ek_pull_request/122/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
The following instances are being terminated: https://elasticbox.com/services/instances/i-u5uxfr

**Comment by allyourbotsarebelongtous** [comment]
Deployed test instance is available at http://ek-43.internal.elastickube.com

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/ek_pull_request/125/ for more details.

**Comment by EfrenRey** [comment]
Jenkins, test this please.

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/ek_pull_request/132/ for more details.

**Comment by osanjuan** [comment]
jenkins test this pleaase

**Comment by allyourbotsarebelongtous** [comment]
Build was ABORTED. See http://jenkins.elasticbox.com/job/ek_pull_request/133/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/ek_pull_request/138/ for more details.

**Comment by EfrenRey** [comment]
Jenkins, test this please

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/ek_pull_request/139/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Deployed test instance is available at http://ek-43.internal.elastickube.com

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/ek_pull_request/142/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
The following instances are being terminated: https://elasticbox.com/services/instances/i-a74kd0
